### PR TITLE
fix(iOS) [0.74]: RCTRedBox not appearing in Bridgeless when metro is not running

### PR DESCRIPTION
**Pull request by okwasniewski**
## Summary:

When testing out `0.74.0-rc0` I found that when the metro is not running we are not displaying RedBox which bumps users to start the packager and reload the app. It also fixes the case where users try to reload by clicking the "Reload" button on RedBox.

## Before

https://github.com/facebook/react-native/assets/52801365/086c557f-ea1f-4a97-b4c7-df8a945cc7a0

## After

https://github.com/facebook/react-native/assets/52801365/9f8421b3-5e83-466f-8cdb-38f97981275d

## Changelog:

[IOS] [FIXED] - RCTRedBox not appearing in Bridgeless when metro is not running

## Test Plan:

Build the app without metro running check if RedBox is shown

### CODE REVIEW

**Review by dmytrorykun — CHANGES_REQUESTED** [review]
It looks like error handling is not implemented in RCTInstance. To solve this properly please implement a separate method called `handleError` that would be basically this: 
https://github.com/facebook/react-native/blob/main/packages/react-native/React/CxxBridge/RCTCxxBridge.mm#L1068
Note that the `_loading` ivar does not exist in RCTInstance. You should assume it `true`.

File: packages/react-native/ReactCommon/react/runtime/platform/ios/ReactCommon/RCTInstance.mm [hunk 1]
```diff
@@ -339,6 +341,10 @@ - (void)_start
     if (RCTGetUseNativeViewConfigsInBridgelessMode()) {
       installLegacyUIManagerConstantsProviderBinding(runtime);
     }
+    
+    RCTExecuteOnMainQueue(^{
+      RCTRegisterReloadCommandListener(self);
```
Comment (RSNara) [inline packages/react-native/ReactCommon/react/runtime/platform/ios/ReactCommon/RCTInstance.mm:346]: I'm not sure if this reload change is necessary! 

But, if it is, I have a request: Can we create a custom implementation of RCTReloadListener that retains RCTInstance weakly?

Bridgeless mode adheres to this principle: 
> Nothing outside of the RCTHost should own a reference to RCTInstance.

In practice, this line is fine: RCTRegisterReloadCommandListener holds on to RCTReloadListeners weakly. 

But, I think bridgeless mode should not depend on this guarantee:
1. Weak retention of listeners is an implementation detail of RCTRegisterReloadCommandListener. If someone changes it in the future, this bridgeless principle could be violated.
2. Bridgeless classes should be responsible for enforcing bridgeless principles.

File: packages/react-native/ReactCommon/react/runtime/platform/ios/ReactCommon/RCTInstance.mm [hunk 2]
```diff
@@ -490,4 +516,8 @@ - (void)_handleJSErrorMap:(facebook::react::MapBuffer)errorMap
                      isFatal:errorMap.getBool(JSErrorHandlerKey::kIsFatal)];
 }
 
+- (void)didReceiveReloadCommand { 
+  [self _loadJSBundle:[self->_bridgeModuleDecorator.bundleManager bundleURL]];
+}
+
```
Comment (RSNara) [inline packages/react-native/ReactCommon/react/runtime/platform/ios/ReactCommon/RCTInstance.mm:522]: @okwasniewski, why did you end up registering the RCTInstance as a reload command listener here?

**Review by RSNara — CHANGES_REQUESTED** [review]
I left a few questions!

File: packages/react-native/ReactCommon/react/runtime/platform/ios/ReactCommon/RCTInstance.mm (hunk 2 above)
Comment (okwasniewski) [reply packages/react-native/ReactCommon/react/runtime/platform/ios/ReactCommon/RCTInstance.mm:522]: When clicking Reload on the redbox nothing was happening

File: packages/react-native/ReactCommon/react/runtime/platform/ios/ReactCommon/RCTInstance.mm (hunk 1 above)
Comment (okwasniewski) [reply packages/react-native/ReactCommon/react/runtime/platform/ios/ReactCommon/RCTInstance.mm]: This was needed to make `Reload` button work in Redbox. I understand your concerns, I've tweaked the code to listen to notification through the notification center so now we don't need to worry about retaining this instance. WDYT about this approach?

Comment (RSNara) [reply packages/react-native/ReactCommon/react/runtime/platform/ios/ReactCommon/RCTInstance.mm]: @okwasniewski, the current solution definitely addresses the instance retention issue.

But, I don't think it's a good idea to rely on a global event emission to communicate between RCTRedBox and RCTInstance. Many apps/environments actually run multiple react instances at the same time. Global event emission makes that configuration on iOS more difficult. And also, more generally: if we can avoid reliance on global things, we probably should. 

Was there any problem with creating a custom RCTReloadListener? Perhaps you noticed an issue, or have a concern, that hadn't crossed my mind.

**Comment by dmytrorykun** [comment]
Hey @okwasniewski , we are generally happy with the changes in this PR changes. But we would like to polish some details. Could you please resubmit this PR for the `main` branch. This way we'll be able to import it to our internal repo, and iterate on it faster. The goal is to land these changes before the 0.75-RC3.

**Comment by okwasniewski** [comment]
Hey @dmytrorykun, I'm going to be on PTO next week so I'm afraid I won't manage to implement custom implementation of RCTReloadListener are you guys okay with iterating on it internally?

**Pull request closed without merging**